sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducerClient.cs

```diff
@@ -509,22 +541,64 @@ public virtual Task<string[]> GetPartitionIdsAsync(CancellationToken cancellatio
         ///
         public virtual async Task CloseAsync(CancellationToken cancellationToken = default)
         {
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
             Closed = true;
 
+            var identifier = GetHashCode().ToString();
+
+            // Attempt to close the active transport producers.  In the event that an exception is encountered,
+            // it should not impact the attempt to close the connection, assuming ownership.
+
+            var transportProducerException = default(Exception);
+
             try
             {
-                await InnerProducer.CloseAsync(cancellationToken).ConfigureAwait(false);
+                await GatewayProducer.CloseAsync(cancellationToken).ConfigureAwait(false);
+
+                var pendingCloses = new List<Task>();
+
+                foreach (var producer in PartitionProducers.Values)
+                {
+                    pendingCloses.Add(producer.CloseAsync(CancellationToken.None));
+                }
+
+                PartitionProducers.Clear();
+                await Task.WhenAll(pendingCloses).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                EventHubsEventSource.Log.ClientCloseError(typeof(EventHubConsumerClient), EventHubName, identifier, ex.Message);
+                transportProducerException = ex;
+            }
+
+            // An exception when closing the connection supersedes one observed when closing the
+            // individual transport clients.
 
+            try
+            {
                 if (OwnsConnection)
                 {
                     await Connection.CloseAsync().ConfigureAwait(false);
                 }
             }
-            catch (Exception ex) when (ex is TaskCanceledException || ex is OperationCanceledException)
+            catch (Exception ex)
             {
-                Closed = InnerProducer.Closed;
+                EventHubsEventSource.Log.ClientCloseError(typeof(EventHubConsumerClient), EventHubName, identifier, ex.Message);
+                transportProducerException = null;
                 throw;
             }
+            finally
+            {
+                EventHubsEventSource.Log.ClientCloseComplete(typeof(EventHubConsumerClient), EventHubName, identifier);
```

Comment:
Don't we have a `ClientCloseStart`? It's not being called in this method.

Reply:
Good catch.   Forgot to bring that over.